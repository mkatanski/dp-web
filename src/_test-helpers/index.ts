export * from "./withResponsiveContext";
export * from "./withThemeProvider";
