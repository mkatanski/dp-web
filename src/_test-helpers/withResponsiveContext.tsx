import React from "react";
import { Context as ResponsiveContext } from "react-responsive";
import { Breakpoints } from "config/breakpoints";

export const withMobileSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.mobile }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withTabletSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.tablet }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.desktop }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withLargeDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.largeDesktop }}>
    {Component}
  </ResponsiveContext.Provider>
);
