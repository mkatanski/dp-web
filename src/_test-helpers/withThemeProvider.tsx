import React from "react";
import { ThemeProvider as DefaultThemeProvider } from "lib/theme/ThemeProvider";
import { StyledTheme, Theme } from "styles/theme/_types";
import { createMuiTheme } from "@material-ui/core";

const testTheme: StyledTheme = {
  primary: "primary",
  secondary: "secondary",
  textPrimary: "textPrimary",
  textSecondary: "textSecondary",
  textInverse: "textInverse"
};

const muiTheme = createMuiTheme({
  direction: "ltr"
});

const theme: Theme = Object.assign(testTheme, muiTheme);

export const withThemeProvider = (Component: React.ReactNode) => (
  <DefaultThemeProvider theme={theme}>{Component}</DefaultThemeProvider>
);
