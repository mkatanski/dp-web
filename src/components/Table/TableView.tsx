import React from "react";
import styled from "styled-components";
import { Table } from "@material-ui/core";
import { omit } from "utils/helpers";

export type TableViewProps = {
  striped?: boolean;
};

export const TableView = styled<React.ComponentType<TableViewProps>>(props => (
  <Table stickyHeader {...omit(props, "striped")} />
))`
  position: absolute;

  th {
    padding: ${({ theme }) => theme.spacing(1, 2)};
  }

  td {
    padding: ${({ theme }) => theme.spacing(2, 2)};
  }

  tr:nth-child(even) {
    background: ${({ theme, striped }) =>
      striped ? theme.palette.grey[50] : "inherit"};
  }
`;

TableView.displayName = "TableView";
