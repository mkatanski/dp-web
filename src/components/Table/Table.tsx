import React from "react";
import { TableView, TableViewProps } from "./TableView";
import {
  TableBody,
  TableHead,
  TableProps as MuiTableProps,
  TableContainer as BareTableContainer
} from "@material-ui/core";
import { omit } from "utils/helpers";

// static components
import { Row } from "./Row";
import { Header } from "./Header";
import { TextCell } from "./TextCell";

import styled from "styled-components";

type TableContainerProps = {
  maxHeight?: number | "auto";
};

export type TableProps = React.PropsWithChildren<
  TableViewProps & MuiTableProps & TableContainerProps
>;

export const TableContainer = styled<React.ComponentType<TableContainerProps>>(
  props => <BareTableContainer {...omit(props, "maxHeight")} />
)`
  width: 100%;
  max-height: ${({ maxHeight }) => maxHeight && `${maxHeight}px`};
`;

export const Table: React.FC<TableProps> & {
  Row: typeof Row;
  Header: typeof Header;
  Body: typeof TableBody;
  Head: typeof TableHead;
  TextCell: typeof TextCell;
} = ({ children, ...rest }: TableProps) => {
  return (
    <TableContainer maxHeight={rest.maxHeight}>
      <TableView data-testid="TableView" {...rest}>
        {children}
      </TableView>
    </TableContainer>
  );
};

Table.Row = Row;
Table.Header = Header;
Table.Body = TableBody;
Table.Head = TableHead;
Table.TextCell = TextCell;
