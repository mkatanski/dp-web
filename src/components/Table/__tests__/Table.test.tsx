/* eslint-disable react/display-name */
import React from "react";
import enzyme from "enzyme";
import "jest-styled-components";

/* Uncomment below if needed */
import TestRenderer from "react-test-renderer";
// import { render, getByTestId } from "@testing-library/react";

import { Table } from "../Table";
import { withThemeProvider } from "_test-helpers";

describe("components", () => {
  describe("Table", () => {
    it("should render TableView", () => {
      const component = enzyme.shallow(<Table />);
      expect(component.find("TableView").exists()).toBe(true);
    });
  });

  it("should render TableView component with maxHeight", () => {
    const tree = TestRenderer.create(
      withThemeProvider(<Table maxHeight={500} />)
    );
    expect(tree).toMatchSnapshot();
  });
});
