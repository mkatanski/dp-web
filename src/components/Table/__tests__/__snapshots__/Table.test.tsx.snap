// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components should render TableView component with maxHeight 1`] = `
.c1 {
  position: absolute;
}

.c1 th {
  padding: 8px 16px;
}

.c1 td {
  padding: 2px 16px;
}

.c1 tr:nth-child(even) {
  background: inherit;
}

.c0 {
  width: 100%;
  max-height: 500px;
}

<div
  className="c0"
  data-props={Object {}}
  id="MUI.TableContainer"
>
  <table
    className="c1"
    data-props={
      Object {
        "data-testid": "TableView",
        "maxHeight": 500,
        "stickyHeader": true,
      }
    }
    id="MUI.Table"
  />
</div>
`;
